cmd/codegen/generator/go/templates/modules.go

```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
 			}
-		} else {
-			argTypeDef, err := ps.goTypeToAPIType(param.Type(), nil)
-			if err != nil {
-				return nil, fmt.Errorf("failed to convert param type: %w", err)
+
+			parent := &paramSpec{
+				name:      params.At(i).Name(),
+				paramType: param.Type(),
+				baseType:  param.Type(),
 			}
 
-			fnDef = dotLine(fnDef, "WithArg").Call(Lit(param.Name()), argTypeDef)
+			for f := 0; f < paramType.NumFields(); f++ {
+				spec, err := ps.parseParamSpecVar(paramType.Field(f))
+				if err != nil {
+					return nil, err
+				}
+				spec.parent = parent
+				spec.description = stype.Fields.List[f].Doc.Text()
+				if spec.description == "" {
+					spec.description = stype.Fields.List[f].Comment.Text()
+				}
+				spec.description = strings.TrimSpace(spec.description)
+				specs = append(specs, spec)
+			}
+			return specs, nil
 		}
 	}
 
-	return fnDef, nil
+	for ; i < params.Len(); i++ {
+		spec, err := ps.parseParamSpecVar(params.At(i))
+		if err != nil {
+			return nil, err
+		}
+		if sig.Variadic() && i == params.Len()-1 {
+			spec.variadic = true
+		}
+
+		pos := fnDecl.Type.Params.List[i].Pos()
+		if cmt, err := ps.commentForPos(pos - 1); err == nil {
+			spec.description = cmt.Text()
+			spec.description = strings.TrimSpace(spec.description)
+		}
+
+		specs = append(specs, spec)
+	}
+	return specs, nil
 }
 
-func namedOrDirectStruct(t types.Type) (*types.Struct, bool) {
-	switch x := t.(type) {
-	case *types.Named:
-		return namedOrDirectStruct(x.Underlying())
-	case *types.Struct:
-		return x, true
-	default:
-		return nil, false
+func (ps *parseState) parseParamSpecVar(field *types.Var) (paramSpec, error) {
+	if _, ok := field.Type().(*types.Struct); ok {
+		return paramSpec{}, fmt.Errorf("nested structs are not supported")
+	}
+
+	paramType := field.Type()
+	baseType := paramType
+	for {
+		ptr, ok := baseType.(*types.Pointer)
+		if !ok {
+			break
+		}
+		baseType = ptr.Elem()
+	}
+
+	optional := false
+	if named, ok := baseType.(*types.Named); ok {
+		if named.Obj().Name() == "Optional" {
+			typeArgs := named.TypeArgs()
+			if typeArgs.Len() != 1 {
+				return paramSpec{}, fmt.Errorf("optional type must have exactly one type argument")
+			}
+			optional = true
+
+			baseType = typeArgs.At(0)
+			for {
+				ptr, ok := baseType.(*types.Pointer)
+				if !ok {
+					break
+				}
+				baseType = ptr.Elem()
+			}
+		}
 	}
+
+	return paramSpec{
+		name:      field.Name(),
+		paramType: paramType,
+		baseType:  baseType,
+		optional:  optional,
+	}, nil
+}
+
+type paramSpec struct {
+	name        string
+	description string
+
+	// paramType is the full type declared in the function signature, which may
+	// include pointer types, Optional, etc
+	paramType types.Type
+	// baseType is the simplified base type derived from the function signature
+	baseType types.Type
+
+	optional bool
+	variadic bool
+
+	defaultValue string
+
+	parent *paramSpec
+}
+
+func (spec *paramSpec) graphqlName() string {
```

Comment:
nit: after https://github.com/dagger/dagger/pull/5932 the engine always provides object+function names with the same name (including casing) that they were originally provided by the SDK as (i.e. SDK registers function with `CamelCase`, engine will invoke SDK runtime using that too even though internally those get converted to `lowerCamelCase` for the gql schema).

So I wouldn't have thought we'd need to handle anything like this anymore. But may be missing something. Not a big deal, just curious if this is necessary anymore.

Reply:
Aha, this isn't for object+function names, this is for the parameter names.

Even if we did the same transformation in the engine, we need to know how to get the graphql input string so we can extract the right name from the incoming request.